components/gateway/src/main/kotlin/net/corda/p2p/gateway/messaging/http/DynamicX509ExtendedTrustManager.kt

```
@@ -0,0 +1,90 @@
+package net.corda.p2p.gateway.messaging.http
+
+import net.corda.p2p.gateway.certificates.RevocationChecker
+import net.corda.p2p.gateway.messaging.RevocationConfig
+import java.net.Socket
+import java.security.KeyStore
+import java.security.cert.CertificateException
+import java.security.cert.X509Certificate
+import javax.net.ssl.CertPathTrustManagerParameters
+import javax.net.ssl.SSLEngine
+import javax.net.ssl.TrustManagerFactory
+import javax.net.ssl.X509ExtendedTrustManager
+
+internal class DynamicX509ExtendedTrustManager(
+    private val trustStoresMap: TrustStoresMap,
+    private val revocationConfig: RevocationConfig,
+    private val trustManagerFactory: TrustManagerFactory = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm())
+): X509ExtendedTrustManager() {
+    private companion object {
+        val wrongUsageMessage = this::class.java.simpleName + " can only be used by the gateway server."
+        const val invalidClientMessage = "None of the possible trust roots were valid for the client certificate. Error(s): "
+    }
+
+    override fun checkClientTrusted(chain: Array<out X509Certificate>?, authType: String?, socket: Socket?) {
+        val exceptionMessages = mutableSetOf<String>()
```

Do we need to keep all the exceptions? If it fails, we can throw an exception without the messages of all the other exception...

---

I think we want this so the user will know what is wrong with the certificate. Otherwise we are throwing away information which might be really useful for debugging purposes. We could consider ignoring the common exception (for the wrong trust root) and just log the others.

---

OK.